optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/variable/listener/support/VariableListenerSupport.java

```diff
@@ -82,11 +82,13 @@ public void linkVariableListeners() {
                     if (nextGlobalOrder <= globalOrder) {
                         nextGlobalOrder = globalOrder + 1;
                     }
-                    VariableListenerNotifiable notifiable = new VariableListenerNotifiable(variableListener, globalOrder);
+                    VariableListenerNotifiable<Solution_> notifiable =
+                            new VariableListenerNotifiable<>(variableListener, globalOrder);
                     for (VariableDescriptor<Solution_> source : shadowVariableDescriptor.getSourceVariableDescriptorList()) {
-                        List<VariableListenerNotifiable> variableNotifiableList = sourceVariableToNotifiableMap.get(source);
+                        List<VariableListenerNotifiable<Solution_>> variableNotifiableList =
+                                sourceVariableToNotifiableMap.get(source);
                         variableNotifiableList.add(notifiable);
-                        List<VariableListenerNotifiable> entityNotifiableList = sourceEntityToNotifiableMap
+                        List<VariableListenerNotifiable<Solution_>> entityNotifiableList = sourceEntityToNotifiableMap
```

Comment:
This is the meat. The class is now created out of the loop and reused; it is immutable, it can be safely shared.